testing/tests/DevExpress.ui.widgets.dataGrid/editing.integration.tests.js

```diff
@@ -2601,6 +2601,42 @@ QUnit.module('Virtual row rendering', baseModuleConfig, () => {
         assert.equal(visibleRows[0].key, 6, 'first visible row key');
         assert.equal($(dataGrid.getRowElement(1, 0)).find('.dx-texteditor').length, 1, 'row has editor');
     });
+
+    // T939449
+    QUnit.test('The virtual row should not be rendered after removing data rows via push API', function(assert) {
+        // arrange
+        const array = [];
+
+        for(let i = 1; i <= 12; i++) {
+            array.push({ id: i });
+        }
+
+        const dataGrid = $('#dataGrid').dxDataGrid({
+            dataSource: array,
+            keyExpr: 'id',
+            editing: {
+                mode: 'row',
+                allowUpdating: true
+            },
+            scrolling: {
+                mode: 'virtual',
+                rowRenderingMode: 'virtual',
+                useNative: false
+            }
+        }).dxDataGrid('instance');
+
+        this.clock.tick();
+        const store = dataGrid.getDataSource().store();
+
+        // act
+        for(let i = 12; i > 10; i--) {
+            store.push([{ type: 'remove', key: i }]);
+        }
```

Comment:
Maybe this will be better?
```
for(let i = 11; i <= 12; i++) {
    store.push([{ type: 'remove', key: i }]);
}
```
or simply
```
store.push([{ type: 'remove', key: 11 }]);
store.push([{ type: 'remove', key: 12 }]);
```